File: 2022/may/11-5-2022/count-sorted-vowel-strings.py
class Solution:
    def countVowelStrings(self, n: int) -> int:
        vowels = ['a', 'e', 'i', 'o', 'u']
        result = []
        buffer = []

        def dfs(index):
            if len(buffer) > n:
                return
            if len(buffer) == n:
                result.append(''.join(buffer[:]))

            for i, val in enumerate(vowels[index:]):
                buffer.append(val)
                dfs(index+i)
                buffer.pop()

        dfs(0)
        return len(result)

print(Solution().countVowelStrings(33))

File: 2022/may/11-5-2022/count-sorted-vowel-strings.rs

pub fn count_vowel_strings(n: i32) -> i32 {
    
    let vowels = vec!['a', 'e', 'i', 'o', 'u'];
    let mut result = vec![];
    dfs(&mut result, &mut vec![], vowels, 0, n);
    result.len()
}

fn dfs(result: &mut Vec<str>, buffer: &mut Vec<str>, vowels: Vec<str>, index: i32, count: i32) {
    if buffer.len() > count {
        return;
    }
    if buffer.len() == count {
        result.push(buffer.join());
    }

    for i in index..vowels.len() {
        buffer.push(vowels[i]);
        dfs(result, buffer, vowels, i, count);
        buffer.pop();
    }
}


fn main() {
    println!("{:?}", count_vowel_strings(2));
}
